Sources/Customization/DefaultEventProcessor.swift

```diff
@@ -0,0 +1,271 @@
+    var observerRevision: NSObjectProtocol?
+    
+    public init(eventHandler: OPTEventHandler = DefaultEventHandler(),
+                batchSize: Int = DefaultValues.batchSize,
+                timerInterval: TimeInterval = DefaultValues.timeInterval,
+                maxQueueSize: Int = DefaultValues.maxQueueSize,
+                backingStore: DataStoreType = .file,
+                dataStoreName: String = "OPTEventQueue") {
+        self.eventHandler = eventHandler
+        self.batchSize = batchSize > 0 ? batchSize : DefaultValues.batchSize
+        self.timerInterval = timerInterval >= 0 ? timerInterval : DefaultValues.timeInterval
+        self.maxQueueSize = maxQueueSize >= 100 ? maxQueueSize : DefaultValues.maxQueueSize
+        
+        self.backingStore = backingStore
+        self.backingStoreName = dataStoreName
+
+        switch backingStore {
+        case .file:
+            self.dataStore = DataStoreQueueStackImpl<EventForDispatch>(queueStackName: "OPTEventQueue",
+                                                                       dataStore: DataStoreFile<[Data]>(storeName: backingStoreName))
+        case .memory:
+            self.dataStore = DataStoreQueueStackImpl<EventForDispatch>(queueStackName: "OPTEventQueue",
+                                                                       dataStore: DataStoreMemory<[Data]>(storeName: backingStoreName))
+        case .userDefaults:
+            self.dataStore = DataStoreQueueStackImpl<EventForDispatch>(queueStackName: "OPTEventQueue",
+                                                                       dataStore: DataStoreUserDefaults())
+        }
+        
+        if self.maxQueueSize < self.batchSize {
+            self.logger.e(.eventDispatcherConfigError("batchSize cannot be bigger than maxQueueSize"))
+            self.maxQueueSize = self.batchSize
+        }
+        
+        addProjectChangeNotificationObservers()
+        
+        subscribe()
+    }
+    
+    deinit {
+        stopTimer()
+        
+        removeProjectChangeNotificationObservers()
+
+        unsubscribe()
+    }
+    
+    open func process(event: BatchEvent, completionHandler: DispatchCompletionHandler? = nil) {
+        guard dataStore.count < maxQueueSize else {
+            let error = OptimizelyError.eventDispatchFailed("EventQueue is full")
+            self.logger.e(error)
+            completionHandler?(.failure(error))
+            return
+        }
+        
+        guard let body = try? JSONEncoder().encode(event) else {
+            let error = OptimizelyError.eventDispatchFailed("Event serialization failed")
+            self.logger.e(error)
+            completionHandler?(.failure(error))
+            return
+        }
+        
+        dataStore.save(item: EventForDispatch(body: body))
+        
+        if dataStore.count >= batchSize {
+            flush()
+        } else {
+            startTimer()
+        }
+        
+        completionHandler?(.success(body))
```

Comment:
Does completion handler make sense here? Nothing is asynchronous.

Reply:
In fact, it's asynchronous. We serialize the event to be stored persistently, which takes a significant delay. That's why we call processEvent asynchronously. We have a couple of cases for process failures, so handler callback will be useful. It's optional.